nucleus/utils.py

```diff
@@ -422,3 +434,87 @@ def paginate_generator(
+    s3_output = s3_prefix.split("//")[-1].split(s3_bucket)[-1][1:]
+    try:
+        boto3.resource("s3").Bucket(s3_bucket).Object(s3_output + '.jpeg').load()
+    except:
+        return None, None
+    try:
+        boto3.resource("s3").Bucket(s3_bucket).Object(s3_output + '.json').load()
+    except:
+        return s3_prefix + '.jpeg', None
+    return s3_prefix + '.jpeg', s3_prefix + '.json'
+
+def upload_chip(s3_prefix: str, chipped_image: Image.Image, chipped_annotations: [dict]) -> tuple[str, str]:
+    s3_bucket = s3_prefix.split("//")[-1].split("/")[0]
+    s3_output = s3_prefix.split("//")[-1].split(s3_bucket)[-1][1:]
+    byteio = io.BytesIO()
+    chipped_image.save(byteio, format="jpeg")
+    byteio.seek(0)
+    try:
+        boto3.resource("s3").Bucket(s3_bucket).Object(s3_output + '.jpeg').upload_fileobj(byteio)
+    except Exception as err:
+        print("hi", err)
+        return None, None
+    if len(chipped_annotations) == 0:
+        return s3_prefix + '.jpeg', None
+    try:
+        boto3.resource("s3").Bucket(s3_bucket).Object(s3_output + '.json').put(Body=json.dumps(chipped_annotations))
+    except Exception as err:
+        print("sup", err)
+        return None, None
+    return s3_prefix + '.jpeg', s3_prefix + '.json'
+
+def generate_offsets(w: int, h: int, chip_size: int, stride_size: int):
+    xs = np.arange(0, w - stride_size, chip_size - stride_size)
+    ys = np.arange(0, h - stride_size, chip_size - stride_size)
+    if len(xs) > 1:
+        xs = np.round(xs * (w - chip_size) / xs[-1]).astype(int)
+    if len(ys) > 1:
+        ys = np.round(ys * (h - chip_size) / ys[-1]).astype(int)
+    yield from product(ys, xs)
+
+def chip_annotations(data: Dict[str,str], x0: int, y0: int, x1: int, y1: int):
+    annotations = list(map(lambda x: {**x["geometry"], "label": x["label"]}, data))
+    if len(annotations):
+        annotations = pd.DataFrame(annotations)[["label", "x", "y", "width", "height"]]
+        annotations["x1"] = annotations["x"] + annotations["width"]
+        annotations["y1"] = annotations["y"] + annotations["height"]
+        annotations["x"] = np.clip(annotations["x"], x0, x1) - x0
+        annotations["x1"] = np.clip(annotations["x1"], x0, x1) - x0
+        annotations["y"] = np.clip(annotations["y"], y0, y1) - y0
+        annotations["y1"] = np.clip(annotations["y1"], y0, y1) - y0
+        annotations["width"] = annotations["x1"] - annotations["x"]
+        annotations["height"] = annotations["y1"] - annotations["y"]
+        annotations["area"] = annotations["width"] * annotations["height"]
+        annotations = annotations.loc[annotations["area"] > 0]
+        annotations = (
+            annotations.groupby(["label", "x", "y", "width", "height"]).last().reset_index()
+        )
+    else:
+        annotations = pd.DataFrame([])
+    annotation_units = []
+    for _, annotation in annotations.iterrows():
+        annotation_units.append({
+            LABEL_KEY: annotation[LABEL_KEY],
```

Comment:
The keys for `annotation` are used and/or set on the lines above and probably should be literals here. For example `LABEL_KEY: annotation["label"]`